libcaf_io/caf/io/network/socket_utils.hpp

```diff
@@ -0,0 +1,165 @@
+/******************************************************************************
+ *                       ____    _    _____                                   *
+ *                      / ___|  / \  |  ___|    C++                           *
+ *                     | |     / _ \ | |_       Actor                         *
+ *                     | |___ / ___ \|  _|      Framework                     *
+ *                      \____/_/   \_|_|                                      *
+ *                                                                            *
+ * Copyright 2011-2018 Dominik Charousset                                     *
+ *                                                                            *
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#pragma once
+
+#include <string>
+
+#include "caf/expected.hpp"
+
+#include "caf/io/network/native_socket.hpp"
+
+#ifdef CAF_WINDOWS
+# ifndef WIN32_LEAN_AND_MEAN
+#   define WIN32_LEAN_AND_MEAN
+# endif // WIN32_LEAN_AND_MEAN
+# ifndef NOMINMAX
+#   define NOMINMAX
+# endif
+# ifdef CAF_MINGW
+#   undef _WIN32_WINNT
+#   undef WINVER
+#   define _WIN32_WINNT WindowsVista
+#   define WINVER WindowsVista
+#   include <w32api.h>
+# endif
+# include <winsock2.h>
+# include <windows.h>
+# include <ws2tcpip.h>
+# include <ws2ipdef.h>
+#else
+# include <unistd.h>
+# include <cerrno>
+# include <sys/socket.h>
+# include <netinet/in.h>
+# include <netinet/ip.h>
+#endif
+
+namespace caf {
+namespace io {
+namespace network {
+
+  // annoying platform-dependent bootstrapping
+#ifdef CAF_WINDOWS
+  using setsockopt_ptr = const char*;
+  using getsockopt_ptr = char*;
+  using socket_send_ptr = const char*;
+  using socket_recv_ptr = char*;
+  using socklen_t = int;
```

Comment:
Defining UNIX types was OK in a `.cpp` file where the symbols wouldn't leak out. However, here we should not "overload" those names.

I propose `socklen_t => socket_size_type` and `ssize_t => signed_size_type`.

Reply:
Do we have to defined these ourselves or which header do we have to include here?
A comment next to a win header reads `// socket_size_type, etc. (MSVC20xx)`, but I wouldn't know where to get them on a unix system.